File: backend/src/models/Food.js
/**
 * Food Model - Enhanced with hotel/admin reference
 */
const mongoose = require('mongoose');

const foodSchema = new mongoose.Schema({
  name: { type: String, required: true, trim: true, maxlength: 100 },
  description: { type: String, required: true, maxlength: 500 },
  price: { type: Number, required: true, min: 0 },
  // Link to admin user who owns this food item
  hotelId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  hotelName: { type: String, required: true, trim: true },
  image: { type: String, default: 'https://images.unsplash.com/photo-1546069901-ba9599a7e63c?w=400' },
  category: { type: String, default: 'General' },
  isAvailable: { type: Boolean, default: true },
  rating: { type: Number, default: 0, min: 0, max: 5 },
  totalRatings: { type: Number, default: 0 },
  preparationTime: { type: Number, default: 20 }, // minutes
  calories: { type: Number },
  isVegetarian: { type: Boolean, default: false },
  isSpicy: { type: Boolean, default: false },
  isFeatured: { type: Boolean, default: false },
  discount: { type: Number, default: 0, min: 0, max: 100 },
  allergens: [{ type: String }],
  ingredients: [{ type: String }],
  sizes: [{
    name: { type: String },
    price: { type: Number }
  }],
  addons: [{
    name: { type: String },
    price: { type: Number }
  }],
  // NEW: Barcode for scanning (unique index defined separately below)
  barcode: { type: String },
  // NEW: Menu types - food can be available for multiple menu types
  menuTypes: [{ 
    type: String, 
    enum: ['delivery', 'dine_in', 'takeaway'], 
    default: ['delivery', 'dine_in'] 
  }],
  // NEW: Dine-in specific pricing (optional, if different from delivery)
  dineInPrice: { type: Number },
  // Engagement metrics
  viewCount: { type: Number, default: 0 },
  likeCount: { type: Number, default: 0 },
  likedBy: [{ type: mongoose.Schema.Types.ObjectId, ref: 'User' }],
  orderCount: { type: Number, default: 0 }
}, { timestamps: true });

// Index for efficient queries
foodSchema.index({ hotelId: 1, category: 1 });
foodSchema.index({ hotelName: 1 });
foodSchema.index({ likeCount: -1 });
foodSchema.index({ viewCount: -1 });
foodSchema.index({ barcode: 1 }, { unique: true, sparse: true }); // NEW: Unique index for barcode lookup
foodSchema.index({ menuTypes: 1 }); // NEW: Index for menu type filtering

module.exports = mongoose.model('Food', foodSchema);

File: frontend/lib/data/models/food.dart
/// Food Model with Hotel Reference
class Food {
  final String id;
  final String name;
  final String description;
  final double price;
  final String hotelId;
  final String hotelName;
  final String image;
  final String category;
  final bool isAvailable;
  final double rating;
  final int totalRatings;
  final int preparationTime;
  final int? calories;
  final bool isVegetarian;
  final bool isSpicy;
  final bool isFeatured;
  final double discount;
  // Hotel info (populated)
  final String? hotelImage;
  final double? hotelRating;
  final String? hotelAddress;
  final bool? hotelIsOpen;
  final double? hotelDeliveryFee;

  Food({
    required this.id,
    required this.name,
    required this.description,
    required this.price,
    required this.hotelId,
    required this.hotelName,
    this.image = '',
    this.category = 'General',
    this.isAvailable = true,
    this.rating = 0,
    this.totalRatings = 0,
    this.preparationTime = 20,
    this.calories,
    this.isVegetarian = false,
    this.isSpicy = false,
    this.isFeatured = false,
    this.discount = 0,
    this.hotelImage,
    this.hotelRating,
    this.hotelAddress,
    this.hotelIsOpen,
    this.hotelDeliveryFee,
  });

  factory Food.fromJson(Map<String, dynamic> json) {
    // Handle populated hotelId
    String hotelId = '';
    String? hotelImage;
    double? hotelRating;
    String? hotelAddress;
    bool? hotelIsOpen;
    double? hotelDeliveryFee;

    if (json['hotelId'] is Map) {
      final hotel = json['hotelId'] as Map<String, dynamic>;
      hotelId = hotel['_id'] ?? hotel['id'] ?? '';
      hotelImage = hotel['hotelImage'];
      hotelRating = (hotel['hotelRating'] as num?)?.toDouble();
      hotelAddress = hotel['hotelAddress'];
      hotelIsOpen = hotel['isOpen'];
      hotelDeliveryFee = (hotel['deliveryFee'] as num?)?.toDouble();
    } else {
      hotelId = json['hotelId']?.toString() ?? '';
    }

    return Food(
      id: json['_id'] ?? json['id'] ?? '',
      name: json['name'] ?? '',
      description: json['description'] ?? '',
      price: (json['price'] ?? 0).toDouble(),
      hotelId: hotelId,
      hotelName: json['hotelName'] ?? json['restaurant'] ?? '',
      image: json['image'] ??
          'https://images.unsplash.com/photo-1546069901-ba9599a7e63c?w=400',
      category: json['category'] ?? 'General',
      isAvailable: json['isAvailable'] ?? true,
      rating: (json['rating'] ?? 0).toDouble(),
      totalRatings: json['totalRatings'] ?? 0,
      preparationTime: json['preparationTime'] ?? 20,
      calories: json['calories'],
      isVegetarian: json['isVegetarian'] ?? false,
      isSpicy: json['isSpicy'] ?? false,
      isFeatured: json['isFeatured'] ?? false,
      discount: (json['discount'] ?? 0).toDouble(),
      hotelImage: hotelImage,
      hotelRating: hotelRating,
      hotelAddress: hotelAddress,
      hotelIsOpen: hotelIsOpen,
      hotelDeliveryFee: hotelDeliveryFee,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'description': description,
      'price': price,
      'image': image,
      'category': category,
      'isAvailable': isAvailable,
      'isVegetarian': isVegetarian,
      'isSpicy': isSpicy,
      'isFeatured': isFeatured,
      'discount': discount,
      'preparationTime': preparationTime,
    };
  }

  double get finalPrice => discount > 0 ? price * (1 - discount / 100) : price;

  // For backward compatibility
  String get restaurant => hotelName;
}
